src/urllib3/connection.py

```diff
@@ -478,125 +483,161 @@ def connect(self) -> None:
-                self.assert_hostname or server_hostname,
-                hostname_checks_common_name,
-            )
-
-        self.is_verified = context.verify_mode == ssl.CERT_REQUIRED or bool(
-            self.assert_fingerprint
-        )
-
-    def _connect_tls_proxy(self, hostname: str, conn: socket.socket) -> "ssl.SSLSocket":
+    def _connect_tls_proxy(self, hostname: str, sock: socket.socket) -> "ssl.SSLSocket":
         """
         Establish a TLS connection to the proxy using the provided SSL context.
         """
-
         proxy_config = cast(
             ProxyConfig, self.proxy_config
         )  # `_connect_tls_proxy` is called when self._is_using_tunnel() is truthy.
         ssl_context = proxy_config.ssl_context
-
-        if ssl_context:
-            # If the user provided a proxy context, we assume CA and client
-            # certificates have already been set
-            return ssl_wrap_socket(
-                sock=conn,
-                server_hostname=hostname,
-                ssl_context=ssl_context,
-            )
-
-        ssl_context = create_proxy_ssl_context(
-            self.ssl_version,
-            self.cert_reqs,
-            self.ca_certs,
-            self.ca_cert_dir,
-            self.ca_cert_data,
-        )
-
-        # If no cert was provided, use only the default options for server
-        # certificate validation
-        return ssl_wrap_socket(
-            sock=conn,
+        ssl_sock, self.proxy_is_verified = _ssl_wrap_socket_and_match_hostname(
+            sock,
+            cert_reqs=self.cert_reqs,
+            ssl_version=self.ssl_version,
+            ssl_minimum_version=self.ssl_minimum_version,
+            ssl_maximum_version=self.ssl_maximum_version,
             ca_certs=self.ca_certs,
             ca_cert_dir=self.ca_cert_dir,
             ca_cert_data=self.ca_cert_data,
             server_hostname=hostname,
             ssl_context=ssl_context,
+            # Features that aren't implemented for proxies yet:
+            assert_fingerprint=None,
+            assert_hostname=None,
+            cert_file=None,
+            key_file=None,
+            key_password=None,
+            tls_in_tls=False,
+        )
+        return ssl_sock  # type: ignore[return-value]
+
+
+def _ssl_wrap_socket_and_match_hostname(
+    sock: socket.socket,
+    *,
+    cert_reqs: Union[None, str, int],
+    ssl_version: Union[None, str, int],
+    ssl_minimum_version: Optional[int],
+    ssl_maximum_version: Optional[int],
+    cert_file: Optional[str],
+    key_file: Optional[str],
+    key_password: Optional[str],
+    ca_certs: Optional[str],
+    ca_cert_dir: Optional[str],
+    ca_cert_data: Union[None, str, bytes],
+    assert_hostname: Union[None, str, "Literal[False]"],
+    assert_fingerprint: Optional[str],
+    server_hostname: Optional[str],
+    ssl_context: Optional["ssl.SSLContext"],
+    tls_in_tls: bool = False,
+) -> Tuple[Union["ssl.SSLSocket", "SSLTransport"], bool]:
```

Comment:
Not a huge fan of returning tuples, should we make it a small struct instead?

Advantages are: it will be easier to extend if we need it, return doesn't depend on ordering of the parameters.

Reply:
Yeah, we should return a namespaced object, even if it's a namedtuple.